64 bit

mwieder opened the pull request:
Here are my changes to date for the 64-bit linux changes. Still compiles and runs properly on 32-bit linux. Runs up to a point (menubar, tool palette, application browser, seems to halt somewhere in "loading plugins") on 64-bit linux mint and then gives a segfault at

Program received signal SIGSEGV, Segmentation fault.
0x000000000057dcbe in MCBitmapEffectBoxBlur::CalculateRows (this=0x69c4720, p_pass=0)
    at ./src/bitmapeffectblur.cpp:735
735                 t_row_buffer[x] = t_pixel_ptr[x] >> 24;

so it's still a work in progress. If I just use the default plugins then it works properly.

In order to get the engine to compile (still desktop only) I had to install the following libraries:
libxft-dev
libxv-dev
libxcursor-dev
X11proto-xinerama-dev
libxinerama-dev
Xinerama

and had to install g++, which didn't come installed by default.
In order to test the engine, I also had to install the ia32lib libraries so that I could install LiveCodeCommunity and then replace the engine.

I now have a post-build script file in /opt/runrev that reads
rm /opt/runrev/livecodecommunity-6.0.0/livecodecommunity.x86
cp /home/mwieder/Revolution/versions/livecode/_build/linux/debug/engine /opt/runrev/livecodecommunity-6.0.0/livecodecommunity.x86

runrevmark committed:
Initial import.

runrevmark committed:
Updated xcconfigs to allow setting of SDKROOT successfully to 10.4 / 10.5 and 10.6 (10.7 untested).
Added more build folders to .gitignore
Fixed compiler errors when compiling with SDKROOT > 10.4

runrevmark committed:
Merge branch 'bugfix-10838' into release-6.0.1

runrevmark committed:
Update submodules.

livecodeian committed:
Remove atl headers from w32 externals

livecodeian committed:
Merge branch 'bugfix-10853' into release-6.0.1

livecodeian committed:
fix intersect detection with semi-transparent images

livecodeian committed:
Fix gif with no repeatcount (i.e. -1) not animating when loaded

livecodeian committed:
Comment previous change

runrevmark committed:
Fixed bug 10837 - literals that look like numbers must eval as 'both' (string and number).

runrevmark committed:
Fixed bug 10844 - incorrect evaluation of threshold parameter, should use ton() rather than ston().

Commit:
added _internal resolve image [id] <id or name> relative to <object reference>

 This command resolves a short id or name of an image as would be used for an icon
 and sets 'it' to the long ID of the image according to the documented rules for resolving
 icons.

 'it' is set empty if the command fails to resolve the image which means it's not on any stack in memory.

livecodeian committed:
Reduce stack load times due to image loading

runrevmark committed:
Merge branch 'bugfix-10828' into release-6.0.1

runrevmark committed:
Merge branch 'bugfix-10843' into release-6.0.1

runrevmark committed:
Merge branch 'bugfix-10827' into release-6.0.1

runrevmark committed:
Merge branch 'bugfix-10837' into release-6.0.1

runrevmark committed:
Merge branch 'bugfix-10844' into release-6.0.1

runrevmark committed:
Update submodules.

Commit:
64-bit safe typedefs

Commit:
64-bit safe typedefs

Commit:
Merge commit 'd6e38f24803032ae18cea20490be189230d2a69b' into feature/_internalResolveImage

Commit:
made style and error handling changes as required

runrevmark committed:
Modifications to Mac Xcode project to support commercial builds.
Added automatic generation of include/revbuild.h based on version file.
Added automatic insertion of versioning into plists.

livecodeian committed:
add global variable revMessageBoxLastObject

implement new global variable revMessageBoxLastObject available within
the development environment

runrevmark committed:
Updated iOS project for support for commercial builds.
Updated mblad for support of Inneractive in commercial builds.

runrevmark committed:
Merge branch 'feature-commercial_integration' into release-6.0.1

Conflicts:
	engine/engine.xcodeproj/project.pbxproj
	rules/Global.xcconfig

runrevmark committed:
Builder updates to support building of both Community and Commercial editions.

livecodeian committed:
add flag to MCObject to indicate script encryption state

runrevmark committed:
Merge pull request #13 from runrevmark/feature-commercial_integration

Builder updates to support building of both Community and Commercial edi...

runrevmark committed:
Merge branch 'feature-commercial_integration' into release-6.0.1

Conflicts:
	ide
	stage.xcodeproj/project.pbxproj

Commit:
fixed core.h int64 one more time

Commit:
fixed core.h int64 one more time

runrevmark committed:
Changes to allow building of Android Community and Commercial engines.

runrevmark committed:
Merge branch 'bugfix-10848' of https://github.com/runrevian/livecode into bugfix-10848

runrevmark committed:
Merge branch 'bugfix-10848' into release-6.0.1

runrevmark committed:
Merge pull request #15 from runrevmark/feature-commercial_integration

Changes to allow building of Android Community and Commercial engines.

runrevmark committed:
Merge branch 'feature-commercial_integration' into release-6.0.1

runrevmark committed:
Added version info generation to android build script.

runrevmark committed:
Merge pull request #16 from runrevmark/feature-commercial_integration

Added version info generation to android build script.

runrevmark committed:
Removed generated file IMarketBillingService.java and updated .gitignore.

runrevmark committed:
Added prebuilt libs for linux server build (crypto, curl and ssl).

runrevmark committed:
Split up Linux makefiles to enable building of commercial and community.

runrevmark committed:
Merge pull request #17 from runrevmark/feature-commercial_integration

Feature commercial integration - linux makefiles updated.

runrevmark committed:
Merge branch 'feature-commercial_integration' into release-6.0.1

runrevmark committed:
Update prebuilt submodule.

runrevmark committed:
Tweaks to ensure derived files all build correctly.

livecodeian committed:
changes to enable dynamic loading of AdModule class if available

runrevmark committed:
Merge pull request #18 from runrevian/feature-commercial_integration

Changes to enable dynamic loading of AdModule class if available

runrevmark committed:
Merge branch 'bugfix-10836' into release-6.0.1

runrevmark committed:
Merge branch 'feature-ide_resolve_image' into release-6.0.1

runrevmark committed:
Minor tweaks to feature-ide_resolve_image.

runrevmark committed:
Updated Windows project files to support building commercial and community.
Updated fields in Windows RC files.

runrevmark committed:
Merge branch 'release-6.0.1' of https://github.com/runrev/livecode into release-6.0.1

runrevmark committed:
Merge pull request #19 from runrevmark/release-6.0.1

Release 6.0.1

runrevmark committed:
Update ide submodule reference.

runrevmark committed:
Merge branch 'release-6.0.1' of https://github.com/runrev/livecode into release-6.0.1

livecodemichael committed:
Updated auto-updater to use new release structure and separated out the resources from installer stack.

livecodemichael committed:
Updated builder to fetch the appropriate resources stack for the installer based on the build type.

runrevmark committed:
Updated ide submodule reference.

runrevmark committed:
Merge pull request #20 from runrevmichael/release-6.0.1

Installer and builder updates

Commit:
fixed MCPath alloc, eliminated more compiler warnings

Commit:
fixed MCPath alloc, eliminated more compiler warnings

ricochet1k commented:
Wow, this patch is awesome. 64bit LiveCode runs a lot faster than the 32bit version on my machine.

I've been trying to debug the segfault. I'm not entirely sure what's going on at this point, but I have been able to find one slightly more obvious symptom of this bug on line 1401 of engine/src/bitmapeffect.cpp. 

(gdb) print src.bounds
$3 = {x = 44, y = 106, width = 377, height = 27}
(gdb) print state.blur_rect
$4 = {x = 44, y = 103, width = 375, height = 27}

The blur rect is outside the bounds of the src image, causing a negative number to be added to the pointer. Does blur_rect need to be clipped somewhere?

mwieder commented:
It's the graphic effects. I've been working away at this for the last hour or so - both bitmapeffect.cpp and bitmapeffectblur.cpp (as well as the associated headers) need tweaking. All the pointers for reading and storing data are defined as int32_t \* instead of intptr_t *, which would be size-agnostic. The callback routines are fighting me now, but I should get them wrangled into shape soon.

runrevmark committed:
Added (temporary) GPL license exception.

runrevmark committed:
Added missing build steps for encode_errors.rev and hash_strings.rev in kernel-development project.

runrevmark committed:
Added missing build step for encode_errors.rev in kernel-server project.

runrevmark committed:
Updated ide submodule.

runrevmark committed:
Fixed issue with mode_development.cpp compiling.
Fixed build order dependencies.
Added auto generation of revbuild.h from 'version'.
Re-enabled building of revBrowser.
Updated submodule references.

runrevmark commented:
I'm not sure those pointers would be the problem - if the vars are specified as specific int widths then that usually means that is what is needed. For example, 'pixels' are 4-bytes (ARGB) whether on 64-bit or 32-bit.

I had a quick look through the code and couldn't see anything immediately obvious that would cause 64-bit issues in bitmapeffect_._...

What is the stacktrace of the crashes you are getting?

ricochet1k commented:
uint32_t\* for pixels is not the problem, the problem is uint32 + 64bit pointers. If you change the type of all stride variables to size_t it stops segfaulting, but the effect doesn't look right.

ricochet1k commented:
I don't think that specific effect code is at fault at this point because if I enable any graphical effects at all I get ugly crap around the button, which looks like the buffer isn't being zeroed out. Drop shadow works correctly, I'm not sure about all the others.

runrevmark commented:
Hmmm - do you have an example of where this change stops the segfault?

The type of the int that is added to a pointer shouldn't affect the amount that is added to a pointer - if you add 4 to a uint32_t \* then it should advance the pointer by 4 \* sizeof(uint32_t) bytes, which is correct for pixel advances.

ricochet1k commented:
At the location I mentioned above, when a negative offset was added to the pointer, it resulted in an invalid pointer that was about 2 orders of magnitude larger than the others, what you would get if you added (uint32_t)-1 to the value. In 32bit that's not a problem because it acts like a normal negative number.

runrevmark commented:
Ah - apologies - I missed the line number reference in your original comment. I understand what you mean now :)

Off the top of my head, I think it is possible for the blur rect to be offset relative to the src rect as a blur adds pixels surrounding it, and this can then be offset (for example the drop-shadow can be anywhere relative to the original image depending on the angle / offset properties).

Sounds like the issue is to do with signedness / unsignedness of the vars rather than their size... I suspect making the stride vars signed would have a similar effect. [ Theory: size_t is 64-bit unsigned, so when doing a multiplication against a 32-bit signed quantity the latter is promoted to 64-bits; thus you are in the same case as if both vars are 32-bit ].

[ Btw, my thoughts here are entirely theoretical - I've not had a chance to setup and test the 64-bit changes yet ]

ricochet1k commented:
Yes, making them signed should do the same thing. The ironic thing is that the stride itself doesn't need to be signed, but the results of using it in calculations do need to be. I don't like C++'s integer semantics.

Actually, I think using ptrdiff_t would be best because it is being used in pointer arithmetic.

runrevmark commented:
Unfortunately I don't think we are going to have time to merge in any of these changes for 6.0.1...

However, I can pull in a feature branch (feature-linux_x64?) into the livecode repo so more than one person can easily submit pull requests against it if that makes it easier (there seem to be a couple of people looking at the issues).

We can then keep this branch ongoing until it's mature enough to merge into the main branch and start producing x64 builds from it.

runrevmark commented on engine/src/fieldhtml.cpp:
```diff
@@ -1426,8 +1429,9 @@ static void import_html_append_unicode_char(import_html_t& ctxt, uint32_t p_code
 	
 	// If the text is currently native (and there is some) or if the styling has changed
 	// then flush.
-	if (!ctxt . is_unicode && ctxt . byte_count > 0 ||
```
The new brackets aren't quite in the right place here - it should be:

``` c++
if ((!ctxt . is_unicode && ctxt . byte_count > 0) ||
    !import_html_equal_style(ctxt . last_used_style, ctxt . styles[ctxt . style_index] . style)
```

runrevmark commented on engine/src/lnxmplayer.cpp:
```diff
@@ -373,7 +374,8 @@ void MPlayer::play ( bool p_play )
 	{
 		if ( m_playing != p_play )
 		{
-			write_command("pause");
```
A better way to handle this might be to check the prototype of 'write_command' - it should probably be taking a 'const char *' rather than a 'char *' (write_command doesn't modify the parameter its passed).

mwieder commented:
Thanks for the comments. Yeah, the parentheses were done pretty quickly to try to cut down on the myriad of compiler warnings. I'd add a bunch of "const"s as well, but that's for later down the road.

mwieder commented:
Urgh. Hit the close button by accident.

Anyway, yes, those pointers aren't the problem. And making the vars signed seems to stop a symptom but the real problems seem to be farther upstream. Applying a graphic effect to an existing control still segfaults. I've gone through the stack trace back as far as MCX11Context::getImage and I'm still seeing source and desination negative values up there.

ricochet1k commented:
What are you doing to get it to segfault? I was able to add all of the graphics effects to a button with no segfaulting.

Commit:
reverted bad paren changes

Commit:
reverted bad paren changes

Commit:
the rest of the backpedaling

Commit:
the rest of the backpedaling

mwieder commented:
Yeah, I can do that as well. But if I load a stack with controls that have graphic effects on a mouseEnter then I get a segfault when the effect is applied.

mwieder commented:
Matt- if you've got graphic effects working without segfaulting (past the initial application of effects) can you post your changes? Did you change the int pointers to signed?

ricochet1k committed:
Use signed stride to avoid breaking pointer math on 64-bit

ricochet1k committed:
Use signed stride to avoid breaking pointer math on 64-bit

ricochet1k commented:
https://github.com/ricochet1k/livecode/commit/de0286204a5d552f674d27f10a789b2c5b8c5a5b

ricochet1k commented:
And I'm not seeing the graphical glitches I was seeing on my other computer. They are working wonderfully.

mwieder commented:
Matt- I incorporated your changes and they work great! Unfortunately you made the changes on your 6.0.1 branch, so pulling it broke compiling here and I had to back out and make your changes by hand. If you can separate those on a new branch you should do a pull request for the 64-bit branch. If not, they're already incorporated into my branch and they'll go in with the next push anyway.

ricochet1k commented:
You should be able to just fetch and cherry-pick the specific commit.

Commit:
Matt Peterson's stride changes for graphic effects

mwieder commented:
Yeah, but I didn't do that because I didn't expect all the files. There were also merge conflicts in two files, so I took the easy way out to test it.

But good work on the stride adjustments. I did something similar but must have messed something up along the way. It's all working fine now. So far.

runrevmark committed:
Make sure any security on standalones is processed (in commercial).

runrevmark committed:
Merge branch 'release-6.0.1' of https://github.com/runrev/livecode into release-6.0.1

livecodeian committed:
Add dependance on version file to revbuild.h target in Makefiles

livecodeian committed:
Merge branch 'bugfix-revbuild-header' into release-6.0.1

livecodeian committed:
fix wrong size used for bitmap when converting ENHMETAFILE

livecodeian committed:
get transformed rep for image when angle == 0 && rect is locked

livecodeian committed:
keep dimensions of image with broken "filename" link when setting the angle

livecodeian committed:
Merge branch 'bugfix-10853' of https://github.com/runrevian/livecode into bugfix-10853

livecodeian committed:
Merge branch 'bugfix-10858' of https://github.com/runrevian/livecode into bugfix-10858

livecodeian committed:
Merge branch 'bugfix-10864' of https://github.com/runrevian/livecode into bugfix-10864

livecodeian committed:
Merge branch 'bugfix-10853' into release-6.0.1

livecodeian committed:
Merge branch 'bugfix-10858' into release-6.0.1

livecodeian committed:
Merge branch 'bugfix-10864' into release-6.0.1

livecodeian committed:
factor out useful environment variables into environment.linux.makefile - update kernel makefiles to use this so we always have the correct path to the cache folder ($CACHE_DIR)

livecodeian committed:
Merge branch 'bugfix-revbuild-header' into release-6.0.1

ricochet1k commented:
mwieder: Then you are not cherry-picking my commit correctly, because that will not have merge conflicts.

I have created my own pull request, mostly because I want to have that commit be linked directly to me.

mwieder commented:
Um... those two things aren't related. Yes, I already said I didn't cherry-pick correctly and the 6.0.1 changes broke my compile, but I've got the files pulled and integrated now. The merge conflicts were because I had previously modified bitmapeffectblur.h and bitmapeffect.cpp, so our changes had to be reconciled first. Everything's pulled and merged properly now.

ricochet1k commented:
Oh, okay, I misunderstood. The only thing that bugs me is that this commit doesn't link to my github account, even though the diffs are exactly the same.

mwieder commented:
Ah... yes, I pulled your commit and it's in my branch but I forgot to push it. Hang on.

Commit:
Merge branch 'master' of git://github.com/ricochet1k/livecode into 64-bit

mwieder commented:
OK - I think it's pushed properly now with your commit properly linked. Let me know if you find otherwise.

ricochet1k commented:
Looks better, thanks.

mwieder commented:
Whew! It took a bit of fast-forwarding the push after resolving the _same_ merge conflict again from the previous push to get that in place. I'm _so_ glad that Ian fixed the 6.0.1 compile bug. My local branches are so much easier to deal with now.

ricochet1k commented:
Yeah, if history has been modified you have to force the push. And I concur on the compile bug, I just wish there was a make clean.

ricochet1k commented:
By the way, interactive rebase (git rebase -i) is a great way to merge and otherwise clean up commits for a pull request.

mwieder commented:
Thanks - I'll have to try the interactive mode next time.

Commit:
agnostic 32 or 64 bit compiles

Commit:
bringing osx and win32 up to date

ricochet1k committed:
Fix bug with "import snapshot" and greatly increase rendering speed on Linux

ricochet1k committed:
Fix bug with "import snapshot" and greatly increase rendering speed on Linux

Commit:
successful osx compile

Commit:
Merge branch 'fix-linux-rendering' into 64-bit

Commit:
Merge branch 'release-6.0.1' into 64-bit

ricochet1k committed:
Fix linker errors for linux libraries

ricochet1k committed:
Add a clean goal and make all the default goal to run when none are specified

Commit:
32-bit compiles on 64-bit machine

Commit:
return 64-bit machine type

livecodemichael committed:
Updated version number to 6.0.1-rc-1 1512.
Updated builder stacks to handle commercial and community builds.

runrevmark committed:
Merge branch 'bugfix-10861' into release-6.0.1

runrevmark committed:
Update submodules.

ricochet1k committed:
Add missing dependencies

ricochet1k committed:
Don't restrict linker output arch and format

ricochet1k committed:
Fix revpdfprinter for 64-bit

ricochet1k committed:
Add missing libs

ricochet1k committed:
libm.a is not built to be included in a library

ricochet1k committed:
libcrypto and the included libz don't like each other, use system libz

mwieder committed:
merged Matt's latest

mwieder committed:
prebuilt libraries

mwieder committed:
make clean

mwieder committed:
Merge branch 'feature-commercial_integration' of git://github.com/runrevmark/livecode-prebuilt into 64-bit

mwieder committed:
win32/64 LLP64 #defines

mwieder committed:
fixed w32 processor and machine

runrevmark commented:
Hi Matt, Mark,

I found the source of the problem with the merge and how to fix it. The issue starts at the 'prebuilt libraries' commit, this seems to be bringing in the contents of the 'prebuilt' repo into the parent folder (i.e. the livecode repo). This commit can't be pulled in as it would cause those objects to then be part of the livecode repo 'forever' and I wanted to avoid that (since they shouldn't be there). Yesterday I couldn't figure out how to work around this.

However, this morning I worked out how to fix the problem... I pulled this branch into a branch starting way-back (just after the initial commit) so it could be 'fast-forwarded'; then unwound the last 4 commits using 'git reset HEAD~1' until I got where I wanted. I then pulled this modified branch into a new branch off master (I've been doing this in my fork of livecode). Since it looks like Matt's makefile changes had been duplicated (commit object wise - probably due to rebasing) rather than merged, I re-pulled his fix-makefiles branch in too and so now we have a much better situation - the changes on this branch, and those on Matt's correctly flow into a merged branch.

Since that, I've been merging in my changes from the linux_x64 branch I made yesterday in my repo and whilst doing so cleaning up some of the Makefile changes I made. You can see the current work in progress there at the linux_x64_2 branch in runrevmark/livecode. I'm hoping to finish those changes soon (along with some changes I've made to the 'deploy' command for standalone building), with the intent of pulling that branch into the main runrev/livecode repo as feature-linux_x64... Hopefully it will almost be ready for merging into develop then :)

Btw, I did cherry-pick some of the changes:
- The warnings in lnxmplayer\* I left as they were previously as they are a side-effect of some memory ownership bugs with regards the C strings the methods affected manipulate and thus I don't want to 'make them go away' until that is fixed properly.
- The changes to Win / OS X with regards 64-bit I left out for now, I'd prefer we dealt with 64-bit-ness of each platform separately after we've got Linux sorted.
- I also changed 'the processor' to return a fixed string based on the target architecture at the point of compile - this means 32-bit intel will always report as 'x86' and 64-bit intel will always report as 'x64'. I felt this was a more consistent way to given concerns over whether 'uname' returns consistent results.
- I reverted the graphics performance changes since I'd prefer that was pulled into develop separately and directly (Matt's already sent a pull-request for this).

[ The only minor wrinkle left is that because I only pulled in changes up to before the 'prebuilt libraries' commit, the github network graph shows a dangling branch in mwieder's repo. If you want to fix this then something like this should work:
    git reset HEAD~1 // repeat until you are at the commit before 'prebuilt libraries'
    git push -f -u origin 64-bit // this will force an update to the remote repo, rewinding the HEAD of 64-bit back a bit
This one of those cases of 'rewriting history' slightly so I don't know whether its a wise thing to do or not. ]

Anyway, any questions, just ask.

Thanks for all your hard work,

Mark.

mwieder commented:
@runrevmark- can we have the 64-bit builds return "the processor" as x86-64 instead?

mwieder commented:
...and thanks - I was about to do a serious rebasing here, but it sounds like you did the heavy lifting and got it sorted out. Part of the problem was that when I pulled Matt's changes the first time he had already integrated the 6.0.1 chain, and that caused my compiles to fail until Ian fixed a bug, so there was a tangled map to deal with.

So at this point should I cancel my pull request? Should you just reject it? Since it appears the changes are integrated I'm not sure what the next step should be.

runrevmark commented:
About x86-64 - I think that is actually what is returned (I'll check) - I misremembered when writing the comments above.

There's no need to cancel or close the pull request - it's been integrated into a branch that's destined to be merged into 'develop', so I'll close it as merged when that occurs :)

runrevmark commented:
These changes have now been merged into 'develop'. Thanks!
